File: backend/app/schemas/topology.py
"""Request/response schemas for the Topology API (bounded graph queries).

Shared by both traversal engines (PostgreSQL recursive CTE and, when
configured and reachable, Neo4j - see ``app/services/topology.py``) so the
route layer and the frontend never need to know which one actually answered.
"""

from __future__ import annotations

import uuid
from typing import Literal

from pydantic import BaseModel, ConfigDict, Field

from app.schemas.relationship import AssetSummaryRead

Direction = Literal["both", "upstream", "downstream"]

MAX_DEPTH = 3
DEFAULT_DEPTH = 1
MAX_NODES = 500
DEFAULT_NODE_CAP = 200


class TopologyNode(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    name: str
    asset_type: str
    environment: str
    criticality: str
    status: str
    is_active: bool
    #: True for the node the query was rooted at.
    is_root: bool = False


class TopologyEdge(BaseModel):
    id: uuid.UUID
    source_asset_id: uuid.UUID
    target_asset_id: uuid.UUID
    relationship_type: str


class SubgraphResponse(BaseModel):
    root_asset_id: uuid.UUID
    nodes: list[TopologyNode]
    edges: list[TopologyEdge]
    depth: int = Field(ge=0, le=MAX_DEPTH)
    direction: Direction
    #: True when the node cap was hit and the result is a partial view.
    truncated: bool
    #: Which engine actually answered - "postgres" or "neo4j". Never exposed as
    #: a claim of Neo4j being canonical; purely diagnostic for the UI.
    source: Literal["postgres", "neo4j"]


class ImpactedAsset(BaseModel):
    asset: AssetSummaryRead
    #: Hop count from the root along propagating relationship types.
    distance: int = Field(ge=1)
    #: Ordered asset ids from the root to this asset (inclusive of both ends).
    path: list[uuid.UUID]


class ImpactResponse(BaseModel):
    root_asset_id: uuid.UUID
    affected_assets: list[ImpactedAsset]
    max_depth: int = Field(ge=0, le=MAX_DEPTH)
    truncated: bool
    source: Literal["postgres", "neo4j"]


class PathResponse(BaseModel):
    source_asset_id: uuid.UUID
    target_asset_id: uuid.UUID
    found: bool
    #: Ordered nodes from source to target (empty when not found).
    nodes: list[TopologyNode]
    edges: list[TopologyEdge]
    length: int = Field(ge=0)
    source_engine: Literal["postgres", "neo4j"]


class GraphHealth(BaseModel):
    """Neo4j status for the system-health surface. Never fails the overall
    platform health - see ``app/services/health.py``."""

    configured: bool
    status: Literal["operational", "unavailable", "not_configured"]
    #: Best-effort node/edge counts from the last successful check, if any.
    detail: str | None = None

File: frontend/src/types/topology.ts
/** Topology query types - mirror `app/schemas/topology.py` exactly. */

import type { RelationshipType } from "./relationship";

export type TopologyDirection = "both" | "upstream" | "downstream";

export interface TopologyNode {
  id: string;
  name: string;
  asset_type: string;
  environment: string;
  criticality: string;
  status: string;
  is_active: boolean;
  is_root: boolean;
}

export interface TopologyEdge {
  id: string;
  source_asset_id: string;
  target_asset_id: string;
  relationship_type: RelationshipType;
}

export interface SubgraphResponse {
  root_asset_id: string;
  nodes: TopologyNode[];
  edges: TopologyEdge[];
  depth: number;
  direction: TopologyDirection;
  truncated: boolean;
  source: "postgres" | "neo4j";
}

export interface ImpactedAsset {
  asset: {
    id: string;
    name: string;
    asset_type: string;
    environment: string;
    criticality: string;
    status: string;
    is_active: boolean;
  };
  distance: number;
  path: string[];
}

export interface ImpactResponse {
  root_asset_id: string;
  affected_assets: ImpactedAsset[];
  max_depth: number;
  truncated: boolean;
  source: "postgres" | "neo4j";
}

export interface PathResponse {
  source_asset_id: string;
  target_asset_id: string;
  found: boolean;
  nodes: TopologyNode[];
  edges: TopologyEdge[];
  length: number;
  source_engine: "postgres" | "neo4j";
}

export interface GraphHealth {
  configured: boolean;
  status: "operational" | "unavailable" | "not_configured";
  detail: string | null;
}

export interface SubgraphParams {
  rootAssetId: string;
  depth?: number;
  direction?: TopologyDirection;
  relationshipType?: RelationshipType[];
  environment?: string;
  criticality?: string;
  status?: string;
  nodeCap?: number;
}

// --- runtime guards ---------------------------------------------------------

function isRecord(v: unknown): v is Record<string, unknown> {
  return typeof v === "object" && v !== null && !Array.isArray(v);
}

function isTopologyNode(v: unknown): v is TopologyNode {
  return (
    isRecord(v) &&
    typeof v.id === "string" &&
    typeof v.name === "string" &&
    typeof v.asset_type === "string" &&
    typeof v.environment === "string" &&
    typeof v.criticality === "string" &&
    typeof v.status === "string" &&
    typeof v.is_active === "boolean" &&
    typeof v.is_root === "boolean"
  );
}

function isTopologyEdge(v: unknown): v is TopologyEdge {
  return (
    isRecord(v) &&
    typeof v.id === "string" &&
    typeof v.source_asset_id === "string" &&
    typeof v.target_asset_id === "string" &&
    typeof v.relationship_type === "string"
  );
}

export function isSubgraphResponse(v: unknown): v is SubgraphResponse {
  return (
    isRecord(v) &&
    typeof v.root_asset_id === "string" &&
    Array.isArray(v.nodes) &&
    v.nodes.every(isTopologyNode) &&
    Array.isArray(v.edges) &&
    v.edges.every(isTopologyEdge) &&
    typeof v.depth === "number" &&
    typeof v.truncated === "boolean"
  );
}

export function isImpactResponse(v: unknown): v is ImpactResponse {
  return (
    isRecord(v) &&
    typeof v.root_asset_id === "string" &&
    Array.isArray(v.affected_assets) &&
    v.affected_assets.every(
      (a) => isRecord(a) && isRecord(a.asset) && typeof a.distance === "number",
    ) &&
    typeof v.truncated === "boolean"
  );
}

export function isPathResponse(v: unknown): v is PathResponse {
  return (
    isRecord(v) &&
    typeof v.found === "boolean" &&
    Array.isArray(v.nodes) &&
    v.nodes.every(isTopologyNode) &&
    Array.isArray(v.edges) &&
    v.edges.every(isTopologyEdge)
  );
}

export function isGraphHealth(v: unknown): v is GraphHealth {
  return (
    isRecord(v) &&
    typeof v.configured === "boolean" &&
    typeof v.status === "string"
  );
}
